compiler/rustc_ast_pretty/src/pprust/state.rs

```diff
@@ -146,47 +146,146 @@ pub fn print_crate<'a>(
     s.s.eof()
 }
 
-/// Should two consecutive tokens be printed with a space between them?
+fn is_punct(tt: &TokenTree) -> bool {
+    matches!(tt, TokenTree::Token(tok, _) if tok.is_punct())
+}
+
+/// Should two consecutive token trees be printed with a space between them?
+///
+/// NOTE: should always be false if both token trees are punctuation, so that
```

Comment:
I don't think this function should return anything in this case, it should rather have `assert!(!is_punct(tt1) || !is_punct(tt2))` at the start instead.

The decision should be made based on `Spacing` in that case, and we should never reach this function.

Reply:
Or the `Spacing`-based decision can be made inside this function, then it will be `if is_punct(tt1) && is_punct(tt2) { ... }` at the start instead of the assert.

Reply:
I would prefer to avoid making any pretty-printing decisions based on `Spacing` in this PR. We can leave those to #114571, which will change how `space_between` is called. I plan to add the `Spacing::Unknown` in that PR, for tokens coming from proc macros. Those will be the cases where `space_between` is used.

With that decided, the current position of the assertion has the advantage that it's only checked in the case where `space_between` returns false.

So I think this is good enough to merge, or do a crater run if you think that is necessary.

Reply:
Ok.
Crater run is needed in any case.